autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py

```diff
@@ -0,0 +1,479 @@
+        """
+        Fits a component by using an input dictionary with pre-requisites
+
+        Args:
+            X (X: Dict[str, Any]): Dependencies needed by current component to perform fit
+            y (Any): not used. To comply with sklearn API
+
+        Returns:
+            A instance of self
+        """
+        # Make sure that the prerequisites are there
+        self.check_requirements(X, y)
+
+        # Setup the logger
+        self.logger = logging.get_logger(X['job_id'])
+
+        fit_function = self._fit
+        if X['use_pynisher']:
+            wall_time_in_s = X['runtime'] if 'runtime' in X else None
+            memory_limit = X['cpu_memory_limit'] if 'cpu_memory_limit' in X else None
+            fit_function = pynisher.enforce_limits(
+                wall_time_in_s=wall_time_in_s,
+                mem_in_mb=memory_limit,
+                logger=self.logger
+            )(self._fit)
+
+        # Call the actual fit function.
+        state_dict = fit_function(
+            X=X,
+            y=y,
+            **kwargs
+        )
+
+        if X['use_pynisher']:
+            # Normally the X[network] is a pointer to the object, so at the
+            # end, when we train using X, the pipeline network is updated for free
+            # If we do multiprocessing (because of pynisher) we have to update
+            # X[network] manually. we do so in a way that every pipeline component
+            # can see this new network -- via an update, not overwrite of the pointer
+            state_dict = state_dict.result
+            X['network'].load_state_dict(state_dict)
+
+        # TODO: when have the optimizer code, the pynisher object might have failed
+        # We should process this function as Failure if so trough fit_function.exit_status
+        return cast(autoPyTorchComponent, self.choice)
+
+    def _fit(self, X: Dict[str, Any], y: Any = None, **kwargs: Any) -> torch.nn.Module:
+        """
+        Fits a component by using an input dictionary with pre-requisites
+
+        Args:
+            X (X: Dict[str, Any]): Dependencies needed by current component to perform fit
+            y (Any): not used. To comply with sklearn API
+
+        Returns:
+            A instance of self
+        """
+
+        # Comply with mypy
+        assert self.choice is not None
+
+        # Setup a Logger and other logging support
+        if 'use_tensorboard_logger' in X and X['use_tensorboard_logger']:
+            self.writer = SummaryWriter(log_dir=X['working_dir'])
+
+        if X["torch_num_threads"] > 0:
+            torch.set_num_threads(X["torch_num_threads"])
+
+        budget_tracker = BudgetTracker(
+            budget_type=X['budget_type'],
+            max_runtime=X['runtime'] if 'runtime' in X else None,
+            max_epochs=X['epochs'] if 'epochs' in X else None,
+        )
+
+        self.choice.prepare(
+            model=X['network'],
+            metrics=[m() for m in get_metrics(X['dataset_properties'])],
```

Comment:
Are all metrics determined by the dataset properties? I would imaging that the user also wants to specify additional metrics, e.g. log_loss, accuracy and balanced accuracy

Reply:
Thanks, I think Ravin added support for this, so I missed to integrate this. I will add it now.